tests/pipeline/test_hooks.py

```diff
@@ -0,0 +1,402 @@
+                chunksize=5,
+                hooks=hooks,
+            )
+            expected_chunks = [
+                tuple(self.trading_days[[-10, -6]]),
+                tuple(self.trading_days[[-5, -1]]),
+            ]
+        else:
+            self.run_pipeline(
+                pipeline=pipeline,
+                start_date=start_date,
+                end_date=end_date,
+                hooks=hooks,
+            )
+            expected_chunks = [(start_date, end_date)]
+
+        expected_loads = set(TrivialFactor.inputs) | {TestingDataSet.bool_col}
+        expected_computes = {
+            TestingDataSet.bool_col.latest,
+            TrivialFactor(),
+            TrivialFactor().rank(),
+            TrivialFactor().rank().zscore(),
+            Everything(),  # Default input for .rank().
+        }
+
+        for h in hooks + [self.global_testing_hook]:
+            self.verify_trace(
+                h.trace,
+                pipeline,
+                pipeline_start_date=start_date,
+                pipeline_end_date=end_date,
+                expected_loads=expected_loads,
+                expected_computes=expected_computes,
+                expected_chunks=expected_chunks,
+            )
+
+    def verify_trace(self,
+                     trace,
+                     pipeline,
+                     pipeline_start_date,
+                     pipeline_end_date,
+                     expected_loads,
+                     expected_computes,
+                     expected_chunks):
+        """Verify a trace of a Pipeline execution.
+        """
+        # First/last calls should bracket the pipeline execution.
+        self.expect_context_pair(trace[0], trace[-1], 'running_pipeline')
+        self.assertEqual(
+            trace[0].args,
+            (pipeline, pipeline_start_date, pipeline_end_date),
+        )
+
+        # Break up the trace into the traces of each chunk.
+        chunk_traces = self.split_by_chunk(trace[1:-1])
+
+        for ctrace, (chunk_start, chunk_end) in zip(chunk_traces,
+                                                    expected_chunks):
+
+            # First call should create the execution plan.
+            self.assertEqual(ctrace[0].method_name, 'on_create_execution_plan')
+
+            # Next call should bracket compute_chunk
+            self.expect_context_pair(ctrace[1], ctrace[-1], 'computing_chunk')
+            self.assertIsInstance(ctrace[1].args[0], ExecutionPlan)
+            self.assertEqual(ctrace[1].args[1:], (chunk_start, chunk_end))
+
+            # Remainder of calls should be loads and computes. These have to
+            # happen in dependency order, but we don't bother to assert that
+            # here. We just make sure that we see each expected load/compute
+            # exactly once.
+            loads_and_computes = ctrace[2:-1]
+            loads = set()
+            computes = set()
+            for enter, exit_ in two_at_a_time(loads_and_computes):
```

Comment:
we can use `toolz.partition(2, loads_and_computes, pad=None)`. The pad is to make the assertion fail on an odd number of entries.

Reply:
I like the name `two_at_a_time` for the use-site here. I updated the definition to just be a wrapper around `toolz.partition`.